agent/agentssh/forward.go

```diff
@@ -62,13 +72,17 @@ func (h *forwardedUnixHandler) HandleSSHRequest(ctx ssh.Context, _ *ssh.Server,
 		addr := reqPayload.SocketPath
 		log = log.With(slog.F("socket_path", addr))
 		log.Debug(ctx, "request begin SSH unix forward")
+
+		key := forwardKey{
+			sessionID: ctx.SessionID(),
+			addr:      addr,
+		}
+
 		h.Lock()
-		_, ok := h.forwards[addr]
+		_, ok := h.forwards[key]
 		h.Unlock()
 		if ok {
-			log.Warn(ctx, "SSH unix forward request for socket path that is already being forwarded (maybe to another client?)",
-				slog.F("socket_path", addr),
-			)
+			log.Warn(ctx, "SSH unix forward request for socket path that is already being forwarded on this session")
 			return false, nil
```

Comment:
Review: Do we want to match OpenSSH even more closely here?

```suggestion
			// Forwarding a socket to the same location twice does not produce
			// an error for OpenSSH, we match that behavior here.
			//
			// Example: ssh -R /tmp/sock:/my/sock -R /tmp/sock:my/other/sock
			//
			// In this case, the second forward will be ignored.
			log.Warn(ctx, "SSH unix forward request for socket path that is already being forwarded on this session, ignoring")
			return true, nil
```

Reply:
If we get more than one call to start forwarding, would we get more than one call to stop? I there could be all kinds of different behaviors depending on the client code, so sticking close to what OpenSSH does on the server side is a good idea.

Reply:
Interestingly, OpenSSH does _not_ seem to send `cancel-streamlocal-forward@openssh.com`. So there's no call to stop, the forwarding simply stops because the session closes. Yet more interestingly, our client does. It does so for both forwarded sockets (i.e. two times), even though they're remote destination is the same. I think both behaviors are fine, and sending `cancel-streamlocal-forward@openssh.com` twice for the same socket also doesn't seem harmful.

One motivation to follow `OpenSSH` here and return `true` is that it will affect the OpenSSH setting of `ExitOnForwardFailure`. If set to `yes`, and we return `false` here, then SSH cannot be established (`Error: remote port forwarding failed for listen path /tmp/echo.sock`).

Why is OpenSSH so lax here? I honestly don't know. It does seem like a potential foot-gun for users, but I'm still in favor of matching OpenSSH.